src/vmm/src/lib.rs

```diff
@@ -307,12 +309,19 @@ pub struct Vmm {
     // Used by Vcpus and devices to initiate teardown; Vmm should never write here.
     vcpus_exit_evt: EventFd,
 
+    // Allocator for guest resrouces
+    resource_allocator: Rc<ResourceAllocator>,
     // Guest VM devices.
     mmio_device_manager: MMIODeviceManager,
     #[cfg(target_arch = "x86_64")]
     pio_device_manager: PortIODeviceManager,
 }
 
+/// SAFETY: safe to implement Sync
+unsafe impl Sync for Vmm {}
+/// SAFETY: safe to implement Send
+unsafe impl Send for Vmm {}
```

Comment:
Damn! I forgot about this. Thanks for catching it.

So, it builds, but clippy complains about it. I assume that's exactly because I sticked these `Rc` inside it. Need to re-think about it. I'm open to suggestions.

Reply:
Oh, interesting, yeah I didn't run clippy! Clippy is complaining because if you put a non `Send + Sync` type into an `Arc<Mutex<...>>`, then that's kinda useless because the `Arc<Mutex<...>>` won't implement `Send + Sync` either in this case, meaning it cannot be shared between threads. But if an `Arc<Mutex<...>>` cannot be shared between threads, then it really does not need to be an `Arc<Mutex<...>>`, it can just be a `Rc<RefCell<...>>`, which will be more performant. But its not _wrong_ or unsafe to use `Arc<Mutex<...>>` anyway. Just kinda... useless. So it's fine to just suppress the clippy warning here.

Now, in our case we need `Arc<Mutex<...>>` because our event manager expects everything passed into it to be `Arc<Mutex<...>>`'d. Now the interesting point is that the event-manager is single threaded, so I don't really see _why_ we should require everything to be `Arc<Mutex<...>>`. A `Rc<RefCell<...>>` should be just as good (and in fact, we _know_ that `Rc<RefCell<...>>` would work, because a `Arc<Mutex<!Send + Sync>>` is just an `Rc<RefCell<...>>` with inefficient `borrow_mut()` method. If the event-manager actually required the additional functionality provided by `Arc<Mutex<...>>` then we would have a compilation failure as soon as we pass in something that's not `Send + Sync`). 

Going one step further, I wonder if we could even get rid of the `Rc<RefCell<...>>` and just have our event loop be directly `dyn MutEventSubscriber` (or even static dispatch through an enum of everything via register onto the loop). I know at the moment we want access to some fields of for example the `Vmm` struct from outside the event loop, but that should be solvable using channels or similar instead :thinking: 

I'm gonna play around with this for a bit

Reply:
Thanks a lot for the dive deep into this. I will adjust.

Reply:
Fixed. Removed the marker traits and disabled the lint on which clippy was complaining, adding a comment explaining why.